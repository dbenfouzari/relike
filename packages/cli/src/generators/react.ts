import { paramCase, snakeCase } from "change-case";
import fs from "fs";
import path from "path";

import { replace, replaceFileName } from "../utils/template";

/**
 * Generates component index file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateIndexFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateIndexFile = (key: string, componentName: string) => {
  const templateIndexPath = path.resolve(__dirname, "../templates/index/index.ts.txt");

  const templateIndex = fs.readFileSync(templateIndexPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateIndexPath, key, componentName)),
    content: replace(templateIndex, key, componentName, paramCase),
  };
};

/**
 * Generates component style file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStyleFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStyleFile = (key: string, componentName: string) => {
  const templateStylePath = path.resolve(__dirname, "../templates/react-style/__name__.module.scss.txt");

  const templateStyle = fs.readFileSync(templateStylePath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStylePath, key, componentName)),
    content: replace(templateStyle, key, componentName, snakeCase),
  };
};

/**
 * Generates component story file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStoryFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStoryFile = (key: string, componentName: string) => {
  const templateStoryPath = path.resolve(__dirname, "../templates/react-story/__name__.stories.tsx.txt");

  const templateStory = fs.readFileSync(templateStoryPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStoryPath, key, componentName)),
    content: replace(templateStory, key, componentName),
  };
};

/**
 * Generates component test file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateTestFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateTestFile = (key: string, componentName: string) => {
  const templateTestPath = path.resolve(__dirname, "../templates/react-test/__name__.test.tsx.txt");

  const templateTest = fs.readFileSync(templateTestPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateTestPath, key, componentName)),
    content: replace(templateTest, key, componentName),
  };
};

/**
 * Generates component file
 *
 * @param   {string}  key           The key to replace
 * @param   {string}  componentName The component name
 * @param   {boolean} hasStyle      Does the component have style
 * @example
 * generateComponentFile("name", name, flags.style);
 * @returns {any}                   File path and content.
 */
export const generateComponentFile = (key: string, componentName: string, hasStyle: boolean) => {
  const templateComponentPath = hasStyle
    ? path.resolve(__dirname, "../templates/react-component/__name__(if:style).tsx.txt")
    : path.resolve(__dirname, "../templates/react-component/__name__.tsx.txt");

  const templateComponent = fs.readFileSync(templateComponentPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateComponentPath, key, componentName)),
    content: replace(templateComponent, key, componentName),
  };
};
