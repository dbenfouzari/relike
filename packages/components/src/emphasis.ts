/**
 * Define an emphasis.
 *
 * Used in buttons for example, to know how to handle different buttons.
 */
export enum Emphasis {
  low = "low",
  medium = "medium",
  high = "high",
}
