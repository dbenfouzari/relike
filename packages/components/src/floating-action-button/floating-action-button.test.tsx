import { render, screen } from "@testing-library/react";

import Color from "../color";
import { Emphasis } from "../emphasis";
import Icons from "../icons";
import FloatingActionButton, { FloatingActionButtonProps } from "./floating-action-button";

/**
 * Set up the tests
 *
 * @param   {Partial<FloatingActionButtonProps>} props The override props
 * @example
 * setup({ emphasis: Emphasis.high })
 * @returns {any}                                      The result
 */
const setup = (props?: Partial<FloatingActionButtonProps>) => {
  const baseProps: FloatingActionButtonProps = {
    icon: Icons.add,
  };

  const mergedProps = { ...baseProps, ...props };

  return render(<FloatingActionButton {...mergedProps} />);
};

describe("Button", () => {
  it("should render successfully with high emphasis", () => {
    setup({ emphasis: Emphasis.high });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully with medium emphasis", () => {
    setup({ emphasis: Emphasis.medium });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully with low emphasis", () => {
    setup({ emphasis: Emphasis.low });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully when extended", () => {
    setup({ icon: Icons.check, label: "See all results" });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully when disabled", () => {
    setup({ disabled: true });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully with low lightness color", () => {
    setup({ color: Color.fromHSLA(100, 50, 5) });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should render successfully with high emphasis and high lightness color", () => {
    setup({ color: Color.fromHSLA(100, 50, 100), emphasis: Emphasis.high });

    expect(screen.getByRole("button")).toMatchSnapshot();
  });
});
