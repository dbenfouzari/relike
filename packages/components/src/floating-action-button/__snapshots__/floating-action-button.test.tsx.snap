// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render successfully when extended 1`] = `
<button
  class="button button__extended button__high buttonHigh-0-2-3 buttonHigh-d5-0-2-9"
>
  <span
    class="jsx-1654271774 icon material-icons icon"
  >
    check
  </span>
  <span
    class="elm-0-2-10 elm-d0-0-2-11"
  >
    See all results
  </span>
</button>
`;

exports[`Button should render successfully with high emphasis 1`] = `
<button
  class="button button__high buttonHigh-0-2-3 buttonHigh-d2-0-2-6"
>
  <span
    class="jsx-1654271774 icon material-icons icon"
  >
    add
  </span>
</button>
`;
